### ISSUE TITLE
Authorization vs. resolver, access to data/model

### ISSUE THREAD

**Issue by mfn** [Folkloreatelier/laravel-graphql#354]
Whilst integrating GraphQL into an existing code base which already uses Laravel-conventions for authorization, e.g. Policy/Policies I found this issue:
- authorization happens before resolving (👍 )
- however authorization also needs access to the model to make the decision
- also: the existing `authorize` method does not contain the (user) $context

Brief example:
```php
class ModelQuery extends Query {

  public function authorize($root, $args) {
    $user = Auth::user();
    $model = Model::findOrFail($args['id']);
    return $user->can('view', $model);
  }

  public function resolve(…) {
    return Model::findOrFail($args['id']);
  }
}
```

In this example I've to query the DB twice to get the model.

I tried doing the authorization inside `resolve()` but then the behaviour changes and I don't get back a 403 HTTP status code but 200 (but I get back an `errors` result).

I read the docs and looked at the examples but I'm not sure what's the recommended way to handle this?

**Comment by mfn** [comment]
Update:
> also: the existing authorize method does not contain the (user) $context

This is actually wrong. Indeed, the signature in `\Folklore\GraphQL\Support\Field::authorize` is only:
```php
    public function authorize($root, $args)
```
however this is part of the resolver of the generic resolver at `\Folklore\GraphQL\Support\Field::getResolver` which in the end gets called from the webonyx GraphQL library via `\GraphQL\Executor\Executor::resolveOrError`:
```php
            return $resolveFn($source, $args, $context, $info);
```
TL;DR: the `$context` (which equals to `\Folklore\GraphQL\GraphQLController::queryContext` = authenticated user) and `ResolveInfo` both are passed to all methods called from Folklores resolver too, which includes `authenticate`, `authorize` and `resolve` => the method signatures are just incomplete => there's already a PR for it => https://github.com/Folkloreatelier/laravel-graphql/pull/287

> In this example I've to query the DB twice to get the model.

I solved this the following way:
- the methods (`authenticate`), `authorize` and `resolve` are just convenience methods called via the default resolver `\Folklore\GraphQL\Support\Field::getResolver`
- the all have the same signature
- The only difference: if either authenticate or authorize returns `false`, a `\Folklore\GraphQL\Error\AuthorizationError` exception is thrown
- all these methods are executed in serial
- which means you can do anything in your `Type`
- I simply did "cache" values on properties of the type I needed access to between `authorize`/`resolve` as not to query the database unnecessarily

**Comment by mfn** [comment]
> I solved this the following way:
>…
> I simply did "cache" values on properties of the type I needed access to between authorize/resolve as not to query the database unnecessarily

Actually, not true anymore. I did refactor all this code on my side and only implemented a single `resolve()` method.

The code performing authorization and retrieving the data and in need of sharing them is no just sequentially executed. Removing this "caching on property" boilerplate was a relieve and made following the logic in the class much simpler.

Although nice on-paper I did find the dedicated `authenticate` and `authorize` methods in the current form not useful.

Maybe if the return value of this methods would be available in the next one.

Or state propagated via the `$context` (but which is currently not possible because in Folklore the context is always the authenticated user unless this is overridden too).
